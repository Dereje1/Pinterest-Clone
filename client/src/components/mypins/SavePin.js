import React from 'react';
import PropTypes from 'prop-types';
import CardActions from '@mui/material/CardActions';
import SaveAltIcon from '@mui/icons-material/SaveAlt';
import IconButton from '@mui/material/IconButton';
import TextField from '@mui/material/TextField';

const SavePin = ({ isImageError, isDescriptionError, savePic }) => {
  let validation;
  if (isImageError && isDescriptionError) {
    validation = {
      text: 'Invalid image and description',
      color: '#f79f9fd9',
    };
  } else if (isImageError) {
    validation = {
      text: 'Invalid image',
      color: '#f79f9fd9',
    };
  } else if (isDescriptionError) {
    validation = {
      text: 'Invalid description',
      color: '#f79f9fd9',
    };
  } else {
    validation = {
      text: 'Save pin',
      color: '#028b02',
    };
  }
  return (
    <CardActions
      disableSpacing
    >
      <IconButton
        aria-label="Pin Image"
        onClick={savePic}
        disabled={isImageError || isDescriptionError}
      >
        <SaveAltIcon style={{ fontSize: '1.5em', color: validation.color }} />
      </IconButton>
      {
        (isImageError || isDescriptionError) && (
          <TextField
            id="pin-create-validation"
            value={validation.text}
            variant="standard"
            sx={{
              marginTop: 2,
              width: '100%',
              input: {
                fontSize: '0.75em',
                color: validation.color,
                fontWeight: 'bold',
                caretColor: 'white',
              },
            }}
            InputProps={{
              disableUnderline: true,
            }}
          />
        )
      }

    </CardActions>
  );
};

export default SavePin;

SavePin.propTypes = {
  isImageError: PropTypes.bool.isRequired,
  isDescriptionError: PropTypes.bool.isRequired,
  savePic: PropTypes.func.isRequired,
};
