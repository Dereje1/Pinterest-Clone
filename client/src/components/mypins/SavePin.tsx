import React, { useEffect, useState } from 'react';
import SaveAltIcon from '@mui/icons-material/SaveAlt';
import Button from '@mui/material/Button';
import FileDownloadOffIcon from '@mui/icons-material/FileDownloadOff';
import RESTcall from '../../crud'; // pin CRUD

interface SavePinProps {
  isImageError: boolean
  isDescriptionError: boolean
  picPreview: string | ArrayBuffer
  isImageLoaded: boolean
  savePic: () => void
}

function SavePin({
  isImageError,
  isDescriptionError,
  savePic,
  isImageLoaded,
  picPreview,
}: SavePinProps) {
  const [isDuplicateError, setIsDuplicateError] = useState(true);
  const [
    testingForDuplicates,
    setTestingForDuplicates,
  ] = useState({ waiting: false, error: false });

  const testForDuplicate = async (picInPreview: string | ArrayBuffer) => {
    setTestingForDuplicates({ waiting: true, error: false });
    try {
      const { duplicateError } = await RESTcall({
        address: '/api/getDuplicateError/',
        method: 'put',
        payload: { picInPreview },
      });
      setTestingForDuplicates({ waiting: false, error: false });
      setIsDuplicateError(duplicateError);
    } catch (error) {
      setTestingForDuplicates({ waiting: false, error: true });
      setIsDuplicateError(true);
    }
  };

  useEffect(() => {
    if (!isImageError && isImageLoaded) {
      testForDuplicate(picPreview);
    }
  }, [isImageError, isImageLoaded]);

  let validation;
  if (!isImageLoaded) {
    validation = {
      text: 'Image loading...',
      color: '#aa9c9cd9',
    };
  } else if (isImageError && isDescriptionError) {
    validation = {
      text: 'Invalid image and description',
      color: '#f79f9fd9',
    };
  } else if (isImageError) {
    validation = {
      text: 'Invalid image',
      color: '#f79f9fd9',
    };
  } else if (isDescriptionError) {
    validation = {
      text: 'Invalid description',
      color: '#f79f9fd9',
    };
  } else if (testingForDuplicates.waiting) {
    validation = {
      text: 'Processing...',
      color: '#aa9c9cd9',
    };
  } else if (testingForDuplicates.error) {
    validation = {
      text: 'Error processing...',
      color: '#f79f9fd9',
    };
  } else if (isDuplicateError) {
    validation = {
      text: 'Image URL already exists',
      color: '#f79f9fd9',
    };
  } else {
    validation = {
      text: 'Save pin',
      color: 'success',
    };
  }
  const isDisabled = isImageError
   || isDescriptionError
    || isDuplicateError
     || !isImageLoaded
      || testingForDuplicates.waiting
       || testingForDuplicates.error;
  return (
    <Button
      variant={isDisabled ? 'text' : 'contained'}
      startIcon={
        isDisabled
          ? <FileDownloadOffIcon sx={{ mr: 1, fontSize: '1.5em' }} />
          : <SaveAltIcon sx={{ mr: 1, fontSize: '1.5em' }} />
      }
      aria-label="Pin Image"
      onClick={savePic}
      disabled={isDisabled}
      color="success"
      sx={{
        '&:disabled': {
          backgroundColor: validation.color,
          color: 'white',
        },
        borderRadius: 0,
        mt: 1,
      }}
    >
      {validation.text}
    </Button>
  );
}

export default SavePin;
