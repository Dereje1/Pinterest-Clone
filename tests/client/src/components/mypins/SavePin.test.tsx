/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SavePin from '../../../../../client/src/components/mypins/SavePin';
import RESTcall from '../../../../../client/src/crud';

jest.mock('../../../../../client/src/crud');
const mockedRESTcall = jest.mocked(RESTcall);

describe('The save pin action button', () => {
  let props: React.ComponentProps<typeof SavePin>;
  beforeEach(() => {
    props = {
      isImageError: true,
      isDescriptionError: true,
      isImageLoaded: true,
      savePic: jest.fn(),
      picPreview: '',
    };
  });

  test('will render for a pending preview that is loading', () => {
    const updatedProps = {
      ...props,
      isImageError: false,
      isDescriptionError: false,
      isImageLoaded: false,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Image loading...');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for an image and description eror', () => {
    const wrapper = shallow(<SavePin {...props} />);
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Invalid image and description');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for an image error only', () => {
    const updatedProps = {
      ...props,
      isDescriptionError: false,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Invalid image');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for a description error only', () => {
    const updatedProps = {
      ...props,
      isImageError: false,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Invalid description');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for a pending duplicate error call', async () => {
    const updatedProps = {
      ...props,
      isImageError: false,
      isDescriptionError: false,
      isImageLoaded: true,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Processing...');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for a failed duplicate error call', async () => {
    mockedRESTcall.mockImplementationOnce(() => Promise.reject());
    const updatedProps = {
      ...props,
      isImageError: false,
      isDescriptionError: false,
      isImageLoaded: true,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    await Promise.resolve();
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Error processing...');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for a duplicate found error', async () => {
    const updatedProps = {
      ...props,
      isDescriptionError: false,
      isImageError: false,
      isImageLoaded: true,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    await Promise.resolve();
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Image URL already exists');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render for no errors and ready to submit', async () => {
    mockedRESTcall.mockImplementationOnce(() => Promise.resolve({ duplicateError: false }));
    const updatedProps = {
      ...props,
      isImageError: false,
      isDescriptionError: false,
      isImageLoaded: true,
    };
    const wrapper = shallow(<SavePin {...updatedProps} />);
    await Promise.resolve();
    const validation = wrapper.find({ 'aria-label': 'Pin Image' });
    expect(validation.text()).toBe('Save pin');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
