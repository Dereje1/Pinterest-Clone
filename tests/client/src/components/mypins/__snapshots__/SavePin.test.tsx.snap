// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The save pin action button will render for a description error only 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#f79f9fd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Invalid description
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for a duplicate found error 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#f79f9fd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Image URL already exists
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for a failed duplicate error call 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#f79f9fd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Error processing...
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for a pending duplicate error call 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#aa9c9cd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Processing...
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for a pending preview that is loading 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#aa9c9cd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Image loading...
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for an image and description eror 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#f79f9fd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Invalid image and description
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for an image error only 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={true}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "#f79f9fd9",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(FileDownloadOffIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Invalid image
</ForwardRef(Fab)>
`;

exports[`The save pin action button will render for no errors and ready to submit 1`] = `
<ForwardRef(Fab)
  aria-label="Pin Image"
  color="success"
  disabled={false}
  onClick={[MockFunction]}
  sx={
    {
      "&:disabled": {
        "backgroundColor": "success",
        "color": "white",
        "mb": 1,
      },
    }
  }
  variant="extended"
>
  <Memo(ForwardRef(SaveAltIcon))
    sx={
      {
        "fontSize": "1.5em",
        "mr": 1,
      }
    }
  />
  Save pin
</ForwardRef(Fab)>
`;
